tools/thanosconvert/thanosconvert.go

```diff
@@ -0,0 +1,195 @@
+package thanosconvert
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/objstore"
+
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+)
+
+// ThanosBlockConverter converts blocks written by Thanos to make them readable by Cortex
+type ThanosBlockConverter struct {
+	logger log.Logger
+	bkt    objstore.Bucket
+	dryRun bool
+}
+
+type PerUserResults struct {
+	FailedBlocks, ConvertedBlocks, UnchangedBlocks []string
+}
+
+func NewPerUserResults() PerUserResults {
+	return PerUserResults{
+		FailedBlocks:    []string{},
+		ConvertedBlocks: []string{},
+		UnchangedBlocks: []string{},
+	}
+}
+
+func (r *PerUserResults) AddFailed(blockID string) {
+	r.FailedBlocks = append(r.FailedBlocks, blockID)
+}
+func (r *PerUserResults) AddConverted(blockID string) {
+	r.ConvertedBlocks = append(r.ConvertedBlocks, blockID)
+}
+
+func (r *PerUserResults) AddUnchanged(blockID string) {
+	r.UnchangedBlocks = append(r.UnchangedBlocks, blockID)
+}
+
+type Results map[string]PerUserResults
+
+// NewThanosBlockConverter creates a ThanosBlockConverter
+func NewThanosBlockConverter(ctx context.Context, cfg bucket.Config, dryRun bool, logger log.Logger, reg prometheus.Registerer) (*ThanosBlockConverter, error) {
+	bkt, err := bucket.NewClient(ctx, cfg, "thanosconvert", logger, reg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ThanosBlockConverter{
+		bkt:    bkt,
+		logger: logger,
+		dryRun: dryRun,
+	}, err
+}
+
+// Run iterates over all blocks from all users in a bucket
+func (c ThanosBlockConverter) Run(ctx context.Context) (Results, error) {
+	results := make(Results)
+
+	// discover users in the bucket
+	var users []string
+	err := c.bkt.Iter(ctx, "", func(o string) error {
+		users = append(users, strings.TrimSuffix(o, "/"))
+		return nil
+	})
+	if err != nil {
+		return results, err
+	}
+	level.Info(c.logger).Log("msg", "Scanned tenants")
+
+	for _, u := range users {
+		r, err := c.convertUser(ctx, u)
+		results[u] = r
+		if err != nil {
+			return results, errors.Wrap(err, fmt.Sprintf("error converting user %s", u))
+		}
+	}
+	return results, nil
+}
+
+func (c ThanosBlockConverter) convertUser(ctx context.Context, user string) (PerUserResults, error) {
+	results := PerUserResults{}
+
+	err := c.bkt.Iter(ctx, user, func(o string) error {
+		// get block ID from path
+
+		parts := strings.Split(o, "/")
```

Comment:
Once you use the `UserBuckerClient`, this should just be a matter of:

```
o = strings.TrimSuffix(o, "/")
blockID, err := ulid.Parse(o)
```

If the parse fail, just skip it (no error) because it means it's not a block. We don't have just blocks in bucket.